# External Data
In this notebook additional data for all cities will be gathered. The focus has been on geographical data and meteorological data. This sort of data have been the focus as it is inhernitly very different from demographic data and it should be fairly independent from who lives where bar climate change is leading to higher temperatures. The hope is that this will revail some interesting correlations with co2 emmisions for cities that the demographic could not.
 
The notebook consist of three parts:

1. **Geographical Data:** In this part various geographical data will created using feature enginering, external API, and external dataset. 
2. **Climate Data:** In this part various cliamte data for each city will be exstracted from Wikipedia using the `BeautifulSoup` library for web scaping. 
3. **Imputing the Final Data:** Finally there is few missing values which will be imputed before that dataset is saved.

In [1]:
# Package import
import requests
import urllib.request
import pandas as pd
from bs4 import BeautifulSoup
import pickle
import numpy as np
import pandas as pd
import warnings
import re

As of first only the columns `City`, `Contry`, `Latitude` and `Longitude` from the processed dataset will be considered as they are needed to fetch the new data. This is loaded in the cell below, where the `head` is printed to ensure correctness.

In [2]:
# Load data
df_geomet = pd.read_csv(os.path.abspath('../data/Processed/Cities_imputed.csv'), index_col=0).loc[:,['City', 'Country', 'Latitude', 'Longitude']]

# Check data
df_geomet.head()

City        Country   Latitude   Longitude
284  Baltimore(MD)  United States  39.290882  -76.610759
9        Melbourne      Australia -37.814218  144.963161
185         Niamey          Niger  13.524834    2.109823
327          Hanoi        Vietnam  21.029450  105.854444
66          Urumqi          China  43.419754   87.319461

## Geographical Data
In this section the following three features will created:

1. The city's distance to equator 
1. The city's elevation above ground
1. The city's distance to the coast 



### Distance to equator

Latitude is measured as a degree between between -90 to 90 from Antarctic to Arctic. Using the radius of the Earth, which is roughly 6371 km, the distance from the equator can be added which is done as a new feature.

In [3]:
df_geomet["Dist_Equator"] = np.abs(df_geomet["Latitude"]*np.pi/180*6371)

### Elevation above ground
It was also desired to use the elevation of the city. For this an api, called *elevation-api* was found. In the free version the resultion was limited to 5000m but we purchased a key to get it down to 90m resolution which is deemed sufficient as it is cities that is worked with:

In [4]:
# It costs me 0.03$ to run this cell so please run with care
def get_elevation(lat, long):
    query = f"https://elevation-api.io/api/elevation?points=({lat},{long})&resolution=90&key=a-a322u40-98wfe6g9a94--JJcG8P-"
    r = requests.get(query).json()  # json object, various ways you can extract value
            # one approach is to use pandas json functionality:
    elevation = r['elevations'][0]['elevation']
    return elevation

df_geomet['Elevation'] = [get_elevation(x,y) for x,y in zip(df_geomet['Latitude'], df_geomet['Longitude'])]

### Distance to coast

Another geological feature that is deemed insteresting is the distance to the coast. Initially it was desired to use an api for this but none were found at a *fair* price. Instead hardcoded data from NASA is used (this data is publicly avialable at: https://oceancolor.gsfc.nasa.gov/docs/distfromcoast/ ). Based on longitude and latitude (rounded to the nearest 0.04 offset by 0.02) the distance to the coast can be loked up. The data is converted to a 2d Pandas dataframe to allow for quicker lookup. The `head` of the resulting dataframe is also shown below:

In [5]:
dist_to_coast_df = pd.pivot_table(pd.read_table('../data/Raw/dist2coast.txt', header = None), index = 0, columns = 1, values = 2)
dist_to_coast_df.head()

1        -89.98   -89.94   -89.90   -89.86   -89.82   -89.78   -89.74  \
0                                                                       
-179.98  1274.3  1269.99  1265.69  1261.39  1257.09  1252.79  1248.49   
-179.94  1274.3  1269.99  1265.69  1261.39  1257.09  1252.79  1248.49   
-179.90  1274.3  1269.99  1265.69  1261.38  1257.08  1252.78  1248.48   
-179.86  1274.3  1269.99  1265.69  1261.38  1257.08  1252.78  1248.48   
-179.82  1274.3  1269.99  1265.68  1261.38  1257.08  1252.77  1248.47   

1         -89.70   -89.66   -89.62  ...    89.62    89.66    89.70    89.74  \
0                                   ...                                       
-179.98  1244.20  1239.90  1235.61  ...  746.661  742.882  739.111  735.348   
-179.94  1244.19  1239.90  1235.60  ...  746.646  742.868  739.099  735.337   
-179.90  1244.18  1239.89  1235.59  ...  746.630  742.854  739.086  735.326   
-179.86  1244.18  1239.88  1235.58  ...  746.614  742.840  739.074  735.315   
-179.82  1244.17  1239.87  1235.58  ...  746.598  742.825  739.061  735.304   

1          89.78    89.82    89.86    89.90    89.94    89.98  
0                                                              
-179.98  731.592  727.844  724.104  720.373  716.650  712.935  
-179.94  731.582  727.836  724.098  720.368  716.647  712.934  
-179.90  731.573  727.829  724.092  720.364  716.644  712.933  
-179.86  731.564  727.821  724.086  720.360  716.642  712.932  
-179.82  731.554  727.813  724.080  720.355  716.639  712.932  

[5 rows x 4500 columns]

Now all longitudes and latitudes has to be rounded to the nearest 0.04. Here it should be noted, that the fractions are offset by 0.02, so it goes xx.02, xx.06$\dots$. A function to round a number to match the NASA data is implemented below:

In [6]:
def round04(x):
    return round(x - ((x % 0.04) - 0.02), 2)

Using this function and the NASA data all distances to the coast can be calculatede and addedadded as it is done below

In [7]:
df_geomet['Dist_to_coast'] = [round(dist_to_coast_df.loc[long,lat]) 
                                for long,lat in zip(
                                    [round04(x) for x in df_geomet.Longitude], 
                                    [round04(x) for x in df_geomet.Latitude]
                                )
                            ]

With these three *geographic* variables added the data is checked to verify it looks correct. A few of the values seen below were cross checked online to ensure correctness.

In [8]:
df_geomet.head()

City        Country   Latitude   Longitude  Dist_Equator  \
284  Baltimore(MD)  United States  39.290882  -76.610759   4368.946697   
9        Melbourne      Australia -37.814218  144.963161   4204.749152   
185         Niamey          Niger  13.524834    2.109823   1503.892925   
327          Hanoi        Vietnam  21.029450  105.854444   2338.368128   
66          Urumqi          China  43.419754   87.319461   4828.056361   

     Elevation  Dist_to_coast  
284        6.0              2  
9         25.0              4  
185      209.0            781  
327       19.0             77  
66      2071.0           2312

## Meteorological Data
In the section the meteorological Data (climate data) for every city will be added. Most cities in the data set have very a informative Wikipedia site which includes information above the climate/weather. That data that is publicaly available to anyone with a internet connection. In fact it turns out almost every city have a dedicated table that describes the cliamte in the city. Below, as an example, a screenshot of the cliamte table for the city of Berlin be seen. 

![title](figures/Berlin_climate.png)

We decided to use Wikipedia data as it was difficult to find an open API (or cheaply priced) which could return climate data about each city whereas Wikipedia is open to everyone. Getting every Cliamte Table turned out to be a much more cumbersome task than first expected. It turned out that though it appear as almost every city has "same" Climate Table like one above they vary slightly and their format is not conscient. This section of the Notebook will explain how the data was scraped and processed such that it could be used as modelling variable. The Webscraping part heabily depends on the `request` and `BeautifulSoup` libraries. The two libraries makes it easy to scrape information from web pages as it is html-based. The first establish a connection to the webpage and the latter handles the scraping. 

This section is composed of three subparts:

1. **Part 2.1 Setting up the Wikipedia links:** Here the Wikipedia links for each city is prepared.
1. **Part 2.2 Extracting the the Climate Data with Beautiful Soup:** Here every climate table like the one above is extracted and stored in a pandas dataframe.
1. **Part 2.3 Processing the Exstraced Climate Data:**: Finally every dataframe is processed such that we can work with data aftterwards.


### Setting up the Wikipedia links
The city names does not match one-to-one with their respective Wikipedia urls. Most can be easily fixed by removing letters inside parethesis and replacing white spaces with underscores (_). E.g. the city Baltimore located in the American state Maryland has the name *Baltimore(MD)* in the data but its Wikipedia url just contains *Baltimore*. Likewise the South African captial *Cape Town* Wikipedia page is found with url *Cape_Town*. These two quick fixes can be achieved using regular expressions that comes with in buildt library `re`. This is done in the cell below:

In [9]:
df_geomet['wiki_link'] = [re.sub(r'\([^)]*\)', '', city).replace(" ", "_") for city in df_geomet.City]
df_geomet.loc[df_geomet['wiki_link'] == 'Birmingham']

City         Country   Latitude  Longitude  Dist_Equator  \
319  Birmingham(AL)   United States  33.570499 -86.765783   3732.869174   
261      Birmingham  United Kingdom  52.479699  -1.902691   5835.476303   

     Elevation  Dist_to_coast   wiki_link  
319      207.0            331  Birmingham  
261      153.0             80  Birmingham

Not all cities wikiepedia URL is fixed with the above regular expressions. Unfortunally quite some cities need a manuel URL change. This have primaily been due to that the city name refers to two cities located next to eachother like *Rotterdam-Hague* as well as the city name exist in multiple countries. This have been fixing by choosing on the cities under the assumption that they are located close to eachother and thus have *very* similar climate and by specifiying the URL for city name that are in multiple countries. The cells below ensure that the *right* wikiepedia link is established.

In [10]:
def update_wiki_link(df, city, city_update):
    if city in df.wiki_link.values:
        df.loc[df['wiki_link'] == city, 'wiki_link'] = city_update
    else:
        print(city)

In [11]:
cities_updates = [
    ('Santa_Cruz', 'Santa_Cruz_(Bolivia)'),
    ('Cordoba' , 'Córdoba,_Argentina'),
    ('San_Francisco_Bay_Area', 'San_Fransisco'),
    ('Tampa-St._Petersburg', 'Saint_Petersburg_(Florida)'),
    ('Memphis', 'Memphis,_Tennessee'),
    ('Denver-Aurora', 'Aurora,_Colorado'),
    ('Rotterdam-Hague', 'Rotterdam'),
    ('Nashville-Davidson', 'Nashville,_Tennessee'),
    ('Cologne-Bonn', 'Cologne'),
    ('Richmond', 'Richmond,_Virginia'),
    ('Rochester', 'Rochester,_New_York'),
    ('Phoenix-Mesa', 'Phoenix_(Arizona)'),
    ('Leon', 'León,_Guanajuato'),
    ('Buffalo', 'Buffalo,_New_York'),
    ('Allentown-Bethlehem', 'Allentown,_Pennsylvania'),
    ('San_Juan', 'San_Juan,_Puerto_Rico'),
    ('Puebla', 'Puebla_(city)'),
    ('San_Jose', 'San_José,_Costa_Rica'),
    ('Dallas-Fort_Worth', 'Fort_Worth,_Texas'),
    ('Seoul-Incheon', 'Seoul'),
    ('Lome', 'Lomé'),
    ('Chihuahua', 'Chihuahua_City'),
    ('Fukuoka-Kitakyushu', 'Fukuoka'),
    ('Davao', 'Davao_City'),
    ('New_York', 'New_York_City'),
    ('Osaka-Kobe-Kyoto', 'Osaka'),
    ('Kansas_City', 'Kansas_City,_Missouri'),
    ('Minneapolis-St._Paul', 'Minneapolis'),
    ('Portland', 'Portland,_Oregon'),
    ('Phoenix_(Arizona)', 'Phoenix,_Arizona'),
    ('Washington', 'Washington,_D.C.'),
    ('Columbus', 'Columbus,_Ohio'),
    ('Salvador', 'Salvador,_Bahia'),
    ('Xi’an', 'Xi%27an'),
    ('Concepcion', 'Concepción,_Chile'),
    ('Providence', 'Providence,_Rhode_Island')
]

# Update based on the above:
for city, city_update in cities_updates:
    update_wiki_link(
        df = df_geomet,
        city = city,
        city_update = city_update
    )

The city name *Birmingham* and *Valencia* appear twice in the dataset. The reason is that there is Burmingham in the United Kingdom as well as in the United States in Alabama. Similar Valencia is a city in Spain as well as a city in Venezuela. The cell below ensures that the appropriate wikipedia link exist for all four cities (two pairs of duplicates).

In [12]:
df_geomet.loc[(df_geomet['wiki_link'] == 'Birmingham') & (df_geomet['Country'] == 'United States'), "wiki_link"] = 'Birmingham,_Alabama'
df_geomet.loc[(df_geomet['wiki_link'] == 'Valencia') & (df_geomet['Country'] == 'United Venezuela'), "wiki_link"] = 'Valencia,_Carabobo'

### Extracting the meteorological data with Beautiful Soup

To extract the cliamte tables three function have been created.

1. `extract_wiki_page`: this function establish a connection to the selected Wikipdia Page. Using BeutifulSoup it exstact the entire page in raw HTML format.
1. `extract_climate_table`: this function process the BeutifulSoup object that contain the entire Wikipdia page of selected City. If a climate table exists on the page it select only this part of the page otherwise it returns `None`.
1. `process_climate_table`: the last function process the selected climate table such that the HTML is converted into a Pandas DataFrame object. That makes it a lot easier to work with in Python.

The three function are defined the cell below.

In [13]:
def extract_wiki_page(wiki_url):
    response = requests.get(wiki_url)
    status_code = response.status_code

    soup = False

    # Only parse if success
    if status_code == 200:
         soup = BeautifulSoup(response.text,"html.parser")

    return soup


def extract_climate_table(soup):
    tables = soup.findAll('table',{"class":"collapsible"})
    
    # Extract the table with the Cliamte Data if it exist
    select_table = None
    for table in tables:
        header = table.findAll(['th'])[0].getText()[0:12].lower()
        
        if 'climate' in header:
            select_table = table
            break
    
    return select_table



def process_climate_table(table):
    # Get table
    trs = table.findAll(['tr'])[1:]

    # Enumerate over all rows
    for i, cell in enumerate(trs):
        th = cell.find_all('th')
        th_data = [col.text.strip('\n') for col in th]  
        td = cell.find_all('td')
        row = [i.text.replace('\n','') for i in td]

        # Get first row as header
        if i == 0:
            columns = th_data
            city_climate_df_tmp = pd.DataFrame(columns=columns)
        
        elif len(th_data) == 0:
            continue
        
        # Populate data
        else:    
            tmp = pd.DataFrame({key: [value] for key, value in zip(columns, th_data+row)})
            city_climate_df_tmp = city_climate_df_tmp.append(tmp)
    
    city_climate_df_tmp.rename(columns={"Month": "Climate_variable"}, inplace = True)

    # Set an index
    city_climate_df_tmp.set_index('Climate_variable', inplace = True)

    return city_climate_df_tmp

Below can the power of the three functions be seen. Here the city of Berlin cliamte data is extracted and returned in the form of Pandas Dataframe. 

In [14]:
# Berlin example
url = 'https://en.wikipedia.org/wiki/' + 'Berlin'
soup = extract_wiki_page(url)
city_climate_table = extract_climate_table(soup)
process_climate_table(city_climate_table)

Jan          Feb         Mar  \
Climate_variable                                                           
Record high °C (°F)                  15.1(59.2)   18.0(64.4)  25.8(78.4)   
Average high °C (°F)                  2.8(37.0)    4.3(39.7)   8.7(47.7)   
Daily mean °C (°F)                    0.1(32.2)    0.9(33.6)   4.3(39.7)   
Average low °C (°F)                  −2.8(27.0)   −2.4(27.7)   0.4(32.7)   
Record low °C (°F)                 −25.3(−13.5)  −22.0(−7.6)  −16.0(3.2)   
Average precipitation mm (inches)    37.2(1.46)   30.1(1.19)  39.3(1.55)   
Mean monthly sunshine hours                57.6         71.5       119.4   
Average ultraviolet index                     1            1           2   

                                          Apr         May         Jun  \
Climate_variable                                                        
Record high °C (°F)                30.8(87.4)  32.7(90.9)  35.4(95.7)   
Average high °C (°F)               14.3(57.7)  19.4(66.9)  22.0(71.6)   
Daily mean °C (°F)                  9.0(48.2)  14.0(57.2)  16.8(62.2)   
Average low °C (°F)                 3.5(38.3)   8.2(46.8)  11.2(52.2)   
Record low °C (°F)                 −7.4(18.7)  −2.8(27.0)   1.3(34.3)   
Average precipitation mm (inches)  33.7(1.33)  52.6(2.07)  60.2(2.37)   
Mean monthly sunshine hours             191.2       229.6       230.0   
Average ultraviolet index                   4           5           6   

                                          Jul          Aug         Sep  \
Climate_variable                                                         
Record high °C (°F)                37.3(99.1)  38.0(100.4)  32.3(90.1)   
Average high °C (°F)               24.6(76.3)   24.2(75.6)  19.3(66.7)   
Daily mean °C (°F)                 19.1(66.4)   18.5(65.3)  14.2(57.6)   
Average low °C (°F)                13.5(56.3)   13.0(55.4)   9.6(49.3)   
Record low °C (°F)                  4.9(40.8)    4.6(40.3)  −0.9(30.4)   
Average precipitation mm (inches)  52.5(2.07)   53.0(2.09)  39.5(1.56)   
Mean monthly sunshine hours             232.4        217.3       162.3   
Average ultraviolet index                   6            5           4   

                                          Oct          Nov           Dec  \
Climate_variable                                                           
Record high °C (°F)                27.7(81.9)   20.4(68.7)    15.6(60.1)   
Average high °C (°F)               13.8(56.8)    7.3(45.1)     3.3(37.9)   
Daily mean °C (°F)                  9.4(48.9)    4.4(39.9)     1.0(33.8)   
Average low °C (°F)                 5.4(41.7)    1.4(34.5)    −1.6(29.1)   
Record low °C (°F)                 −7.7(18.1)  −12.0(10.4)  −24.0(−11.2)   
Average precipitation mm (inches)  32.2(1.27)   37.8(1.49)    46.1(1.81)   
Mean monthly sunshine hours             114.7         54.9          46.9   
Average ultraviolet index                   2            1             0   

                                           Year  
Climate_variable                                 
Record high °C (°F)                 38.0(100.4)  
Average high °C (°F)                 13.7(56.7)  
Daily mean °C (°F)                    9.3(48.7)  
Average low °C (°F)                   5.0(41.0)  
Record low °C (°F)                 −25.3(−13.5)  
Average precipitation mm (inches)  515.2(20.28)  
Mean monthly sunshine hours             1,727.6  
Average ultraviolet index                     3

#### Extracting all Cities' Climate Data
Now all cities will be processed. Each dataframe will be stored in a dictionay where the key is the Wikipeida enterpeation of the cities name. Be aware that it takes a bit of time to run the following code due to the request calls.

In [15]:
# Initialize dict
climate_data_dict = dict()

for wiki_link in df_geomet.wiki_link:
    url = 'https://en.wikipedia.org/wiki/' + wiki_link
    soup = extract_wiki_page(url)

    if soup:
        city_climate_table = extract_climate_table(soup)

        # If the climate table is available on Wikipedia
        if city_climate_table:
            df_city_climate = process_climate_table(city_climate_table)
            climate_data_dict[wiki_link] = df_city_climate

        # If the wikipedia page does not have climate table
        else: 
            climate_data_dict[wiki_link] = None
            print('No climate data available for city: ', wiki_link)
    else:
        print('Could not find citys: ', wiki_link, 'Wikipedia page')

No climate data available for city:  Cebu
No climate data available for city:  Caracas
No climate data available for city:  Santo_Domingo
No climate data available for city:  Monterrey


As can be seen all cities cliamte table have been exstracted except the cities of `Cebu`, `Caracas`, `Santo_Domingo`, and `Monterrey`. The city of Cebu simply does not have a Climate Table so there is nothing to exstract. However the three remaining cities *does* have a wikipedia page with such climate table. However the format is much different from the other cities so it does not work. Rather than tweeking things for hours we decided to let the cities be. Instead all values will be imputed in the end of the notebook.

As each each city takes a few seconds to scrape, primarily due to the repsonse time of wikipedia, the dictionary that contains a Pandas dataframe with cliamte data for each have been saved. Here the library `pickle` habe been used. “Pickling” is the process whereby a Python object is converted into a byte stream, and “unpickling” is the inverse operation, whereby a byte stream (from a binary file or bytes-like object) is converted back into a python object again.

In [16]:
# Save the climate data in pickle as it take quite some time to geneate.
with open('../Data/processed/climate_data_dict.pickle', 'wb') as handle:
    pickle.dump(climate_data_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)

### Processing the exstraced Climate Data
As mentioned each Wikipedia Page climate page, though they look similar, are different.  This is easily proven by examining two different cities. In the cells below are the dataframe for city of `Austin` and `Freetown` showed. Notice the different names of the climate variables. One prefers Farhenheit over Celcius. One mentions snowfall whereas the other does not. This means that some processing have to be carried out on the data to get it alligned for all cities.

In [17]:
with open('../Data/interim/climate_data_dict.pickle', 'rb') as handle:
    climate_data_dict = pickle.load(handle)
climate_data_dict['Austin']

Jan         Feb         Mar  \
Climate_variable                                                             
Record high °F (°C)                         90(32)      99(37)      98(37)   
Mean maximum °F (°C)                    79.3(26.3)  83.3(28.5)  86.6(30.3)   
Average high °F (°C)                    61.5(16.4)  65.2(18.4)  72.2(22.3)   
Average low °F (°C)                      41.5(5.3)   44.8(7.1)  51.3(10.7)   
Mean minimum °F (°C)                    26.7(−2.9)  28.5(−1.9)   34.7(1.5)   
Record low °F (°C)                         −2(−19)     −1(−18)      18(−8)   
Average precipitation inches (mm)         2.22(56)    2.02(51)    2.76(70)   
Average snowfall inches (cm)              0.4(1.0)   0.2(0.51)        0(0)   
Average precipitation days (≥ 0.01 in)         7.4         7.4         9.2   
Average snowy days (≥ 0.1 in)                  0.3         0.2           0   
Average relative humidity (%)                 67.2        66.0        64.2   
Average dew point °F (°C)                36.1(2.3)   39.6(4.2)   46.2(7.9)   
Mean monthly sunshine hours                  163.8       169.3       205.9   
Percent possible sunshine                       51          54          55   
Average ultraviolet index                        4           6           8   

                                               Apr         May         Jun  \
Climate_variable                                                             
Record high °F (°C)                         99(37)     104(40)     109(43)   
Mean maximum °F (°C)                    91.7(33.2)  95.4(35.2)  98.1(36.7)   
Average high °F (°C)                    79.8(26.6)  86.5(30.3)  92.1(33.4)   
Average low °F (°C)                     58.6(14.8)  66.7(19.3)  72.3(22.4)   
Mean minimum °F (°C)                     42.9(6.1)  54.1(12.3)  64.2(17.9)   
Record low °F (°C)                          30(−1)       40(4)      51(11)   
Average precipitation inches (mm)         2.09(53)   4.44(113)   4.33(110)   
Average snowfall inches (cm)                  0(0)        0(0)        0(0)   
Average precipitation days (≥ 0.01 in)         7.1         8.9         7.7   
Average snowy days (≥ 0.1 in)                    0           0           0   
Average relative humidity (%)                 66.4        71.4        69.5   
Average dew point °F (°C)               55.0(12.8)  63.3(17.4)  68.2(20.1)   
Mean monthly sunshine hours                  205.8       227.1       285.5   
Percent possible sunshine                       53          54          68   
Average ultraviolet index                        9          10          11   

                                                Jul          Aug         Sep  \
Climate_variable                                                               
Record high °F (°C)                         109(43)      112(44)     112(44)   
Mean maximum °F (°C)                    101.0(38.3)  102.9(39.4)  99.1(37.3)   
Average high °F (°C)                     95.6(35.3)   97.0(36.1)  90.5(32.5)   
Average low °F (°C)                      74.4(23.6)   74.6(23.7)  69.4(20.8)   
Mean minimum °F (°C)                     69.9(21.1)   69.0(20.6)  56.9(13.8)   
Record low °F (°C)                           57(14)       58(14)       41(5)   
Average precipitation inches (mm)          1.88(48)     2.35(60)    2.99(76)   
Average snowfall inches (cm)                   0(0)         0(0)        0(0)   
Average precipitation days (≥ 0.01 in)          5.4          4.9         6.7   
Average snowy days (≥ 0.1 in)                     0            0           0   
Average relative humidity (%)                  65.1         63.8        68.4   
Average dew point °F (°C)                68.9(20.5)   68.4(20.2)  65.5(18.6)   
Mean monthly sunshine hours                   317.2        297.9       233.8   
Percent possible sunshine                        74           73          63   
Average ultraviolet index                        11           10           9   

                                

In [18]:
climate_data_dict['Freetown']

Jan          Feb  \
Climate_variable                                                        
Record high °C (°F)                           36.0(96.8)  38.0(100.4)   
Average high °C (°F)                          29.9(85.8)   30.3(86.5)   
Daily mean °C (°F)                            27.3(81.1)   27.6(81.7)   
Average low °C (°F)                           23.8(74.8)   24.0(75.2)   
Record low °C (°F)                            15.0(59.0)   17.8(64.0)   
Average rainfall mm (inches)                   8.0(0.31)    6.0(0.24)   
Average rainy days                                     0            0   
Average relative humidity (%) (at 15:00 LST)          66           67   
Mean monthly sunshine hours                        226.3        215.6   

                                                      Mar          Apr  \
Climate_variable                                                         
Record high °C (°F)                           38.0(100.4)  38.5(101.3)   
Average high °C (°F)                           30.9(87.6)   31.2(88.2)   
Daily mean °C (°F)                             28.1(82.6)   28.5(83.3)   
Average low °C (°F)                            24.4(75.9)   24.8(76.6)   
Record low °C (°F)                             19.4(66.9)   20.0(68.0)   
Average rainfall mm (inches)                   28.0(1.10)   68.0(2.68)   
Average rainy days                                      1            4   
Average relative humidity (%) (at 15:00 LST)           67           68   
Mean monthly sunshine hours                         232.5        207.0   

                                                      May           Jun  \
Climate_variable                                                          
Record high °C (°F)                            35.0(95.0)    35.0(95.0)   
Average high °C (°F)                           30.9(87.6)    30.1(86.2)   
Daily mean °C (°F)                             28.3(82.9)    27.4(81.3)   
Average low °C (°F)                            24.4(75.9)    23.6(74.5)   
Record low °C (°F)                             20.6(69.1)    20.0(68.0)   
Average rainfall mm (inches)                  214.0(8.43)  522.0(20.55)   
Average rainy days                                     15            22   
Average relative humidity (%) (at 15:00 LST)           73            76   
Mean monthly sunshine hours                         189.1         153.0   

                                                       Jul           Aug  \
Climate_variable                                                           
Record high °C (°F)                             33.0(91.4)    32.0(89.6)   
Average high °C (°F)                            28.7(83.7)    28.4(83.1)   
Daily mean °C (°F)                              26.3(79.3)    25.9(78.6)   
Average low °C (°F)                             23.1(73.6)    23.0(73.4)   
Record low °C (°F)                              19.4(66.9)    19.4(66.9)   
Average rainfall mm (inches)                  1,190(46.85)  1,078(42.44)   
Average rainy days                                      27            27   
Average relative humidity (%) (at 15:00 LST)            81            82   
Mean monthly sunshine hours                          102.3          86.8   

                                                       Sep           Oct  \
Climate_variable                                                           
Record high °C (°F)                             31.0(87.8)    35.0(95.0)   
Average high °C (°F)                            29.0(84.2)    29.9(85.8)   
Daily mean °C (°F)                              26.4(79.5)    27.1(80.8)   
Average low °C (°F)                             23.1(73.6)    23.4(74.1)   
Record low °C (°F)                              20.0(68.0)    19.4(66.9)   
Average rainfall mm (inches)                  800.0(31.50)  333.0(13.11)   
Average rainy days                                      24            21   
Average relative humidity (%) (at 15:00 LST)            80            78   
Mean monthl

#### Selecting common Wikipedia Cliamte Variables

More general it can be examined how different each table is by considering the number of variables. In the below cell is every cliamte variable for every table exstracted in one long list. By examining the list as set it can be explored just how many different cliamte variable there are.

In [19]:
city_climate_list = [list(city_climate_df.index.values) 
                    for city, city_climate_df in climate_data_dict.items() 
                    if city_climate_df is not None
                    ]

# Get unique cliamte variables
climate_variables = [city for sublist in city_climate_list for city in sublist]
set_of_climate_variable = set(climate_variables)

# Print results
print('In the cliamte tables there is, {} different names for cliamte variables that is measured.'.format(len(set_of_climate_variable)))

In the cliamte tables there is, 69 different names for cliamte variables that is measured.


Though there are 69 different names for cliamte variables some appear much more frequently and some have different names for the same thing. In the cell below is number of time the variables is seen displayed next the name of the climate variable to indicate which appear the most.

In [20]:
varaible_count = pd.Series(climate_variables).value_counts()
[(count, name) for name, count in zip(varaible_count.index, varaible_count)]

[(297, 'Mean monthly sunshine hours'),
 (270, 'Average low °C (°F)'),
 (270, 'Average high °C (°F)'),
 (254, 'Record high °C (°F)'),
 (253, 'Record low °C (°F)'),
 (252, 'Average relative humidity (%)'),
 (251, 'Daily mean °C (°F)'),
 (191, 'Average precipitation mm (inches)'),
 (102, 'Percent possible sunshine'),
 (85, 'Average rainfall mm (inches)'),
 (77, 'Average ultraviolet index'),
 (62, 'Average rainy days'),
 (57, 'Average snowy days'),
 (56, 'Average precipitation days (≥ 0.1 mm)'),
 (56, 'Average low °F (°C)'),
 (56, 'Average high °F (°C)'),
 (55, 'Record high °F (°C)'),
 (55, 'Record low °F (°C)'),
 (53, 'Mean maximum °F (°C)'),
 (53, 'Mean minimum °F (°C)'),
 (49, 'Average precipitation inches (mm)'),
 (46, 'Mean daily sunshine hours'),
 (45, 'Average precipitation days (≥ 0.01 in)'),
 (42, 'Average precipitation days (≥ 1.0 mm)'),
 (39, 'Average snowfall inches (cm)'),
 (38, 'Average snowy days (≥ 0.1 in)'),
 (33, 'Average dew point °F (°C)'),
 (28, 'Average precipitation 

It can be seen of the 69 variables many represents the same thing - just with a different name. Moreover there are some that just have different units.

It appears as the following five variable 

1. `'Mean monthly sunshine hours'`
2. `'Average high °C'`
3. `'Average low °C'`
4. `'Average precipitation mm'`
5. `'Average precipitation days'`

can be extracted accross almost all cliamte tables if one consider multiple names for the above variables and accept different thresholds for a day to be considered having precipitation. Below a dictionary have been created manually which helps to rename the variable such that they are only called the five above things. 

In [21]:
rename_column = {
    'Mean monthly sunshine hours'               : 'Mean monthly sunshine hours',
    'Average high °C (°F)'                      : 'Average high °C',
    'Average high °F (°C)'                      : 'Average high °C',
    'Average low °C (°F)'                       : 'Average low °C',
    'Average low °F (°C)'                       : 'Average low °C',
    'Average rainfall inches (mm)'              : 'Average precipitation mm',
    'Average rainfall mm (inches)'              : 'Average precipitation mm',
    'Average precipitation inches (mm)'         : 'Average precipitation mm',
    'Average precipitation mm (inches)'         : 'Average precipitation mm',
    'Average precipitation mm'                  : 'Average precipitation mm',
    'Average precipitation days'                : 'Average precipitation days',
    'Average precipitation days (≥ 0.005 inch)' : 'Average precipitation days',
    'Average precipitation days (≥ 0.01 in)'    : 'Average precipitation days',
    'Average precipitation days (≥ 0.01 in.)'   : 'Average precipitation days',
    'Average precipitation days (≥ 0.01 inch)'  : 'Average precipitation days',
    'Average precipitation days (≥ 0.01\xa0mm)' : 'Average precipitation days',
    'Average precipitation days (≥ 0.1 mm)'     : 'Average precipitation days',
    'Average precipitation days (≥ 0.1\xa0mm)'  : 'Average precipitation days',
    'Average precipitation days (≥ 0.2 mm)'     : 'Average precipitation days',
    'Average precipitation days (≥ 0.2\xa0mm)'  : 'Average precipitation days',
    'Average precipitation days (≥ 0.5 mm)'     : 'Average precipitation days',
    'Average precipitation days (≥ 1.0 mm)'     : 'Average precipitation days',
    'Average precipitation days (≥ 1\xa0mm)'    : 'Average precipitation days',
    'Average rainy days'                        : 'Average precipitation days',
    'Average rainy days (≥ 0.01 in)'            : 'Average precipitation days',
    'Average rainy days (≥ 0.01 inch)'          : 'Average precipitation days',
    'Average rainy days (≥ 0.01 mm)'            : 'Average precipitation days',
    'Average rainy days (≥ 0.1 mm)'             : 'Average precipitation days',
    'Average rainy days (≥ 0.2 mm)'             : 'Average precipitation days',
    'Average rainy days (≥ 0.5 mm)'             : 'Average precipitation days',
    'Average rainy days (≥ 1 mm)'               : 'Average precipitation days',
    'Average rainy days (≥ 1.0 mm)'             : 'Average precipitation days',
    'Average rainy days (≥ 1.0\xa0mm)'          : 'Average precipitation days',
    'Average rainy days (≥ 1mm)'                : 'Average precipitation days',
}

In the cell below the function `keep_selected_varibles` has been developed that takes the above dictionary as an input as well as dataframe decribing a city climate. The function only returns the values of the keys of the dictionary. That means that only five varaibles will be considered. An example of the function can be seen below with city of Austin.

In [22]:
def keep_selected_varibles(city_climate_df, rename_column):
    
    # Rename remaing climate variables:
    city_climate_variables = city_climate_df.index.tolist()
    name_change  = {climate_variable : rename_column[climate_variable] for climate_variable  in city_climate_variables if climate_variable in rename_column}
    city_climate_df.rename(name_change, inplace = True)

    # only keep the variables where the name was updated:
    city_climate_df_update = city_climate_df.loc[name_change.values()]

    # Remove duplicate indexes...
    # This can happen if a city e.g. Toronto has 'Average precipitation mm (inches)',
    # 'Average rainfall mm (inches)', 'Average snowfall cm (inches)' i.e. precipitation = rainfall + snowfall
    # THis method keeps the first index. Hopefully in all cases this it the total precipitation.
    city_climate_df_update = city_climate_df_update[~city_climate_df_update.index.duplicated(keep='first')]

    return city_climate_df_update

In [23]:
austin_tmp = climate_data_dict['Austin'].copy()
austin_tmp = keep_selected_varibles(austin_tmp, rename_column)
austin_tmp

Jan         Feb         Mar         Apr  \
Climate_variable                                                              
Average high °C              61.5(16.4)  65.2(18.4)  72.2(22.3)  79.8(26.6)   
Average low °C                41.5(5.3)   44.8(7.1)  51.3(10.7)  58.6(14.8)   
Average precipitation mm       2.22(56)    2.02(51)    2.76(70)    2.09(53)   
Average precipitation days          7.4         7.4         9.2         7.1   
Mean monthly sunshine hours       163.8       169.3       205.9       205.8   

                                    May         Jun         Jul         Aug  \
Climate_variable                                                              
Average high °C              86.5(30.3)  92.1(33.4)  95.6(35.3)  97.0(36.1)   
Average low °C               66.7(19.3)  72.3(22.4)  74.4(23.6)  74.6(23.7)   
Average precipitation mm      4.44(113)   4.33(110)    1.88(48)    2.35(60)   
Average precipitation days          8.9         7.7         5.4         4.9   
Mean monthly sunshine hours       227.1       285.5       317.2       297.9   

                                    Sep         Oct         Nov         Dec  \
Climate_variable                                                              
Average high °C              90.5(32.5)  81.8(27.7)  71.4(21.9)  62.7(17.1)   
Average low °C               69.4(20.8)  60.6(15.9)  50.6(10.3)   42.3(5.7)   
Average precipitation mm       2.99(76)    3.88(99)    2.96(75)    2.40(61)   
Average precipitation days          6.7         7.5         7.5         7.8   
Mean monthly sunshine hours       233.8       215.6       168.3       153.5   

                                   Year  
Climate_variable                         
Average high °C              79.8(26.6)  
Average low °C               59.0(15.0)  
Average precipitation mm     34.32(872)  
Average precipitation days         87.5  
Mean monthly sunshine hours     2,643.7

### Selecting metric represenation of Wikipedia Cliamte Variables
In the above dataframe, that shows the selected cliamte data for Austin, there is both the metric and imperial represenation in a string data format. This needs to be processed such that it is numeric format hence there can only be one represenation. The metric one has been selected. To ensure a numeric metric entepreation of the dataframe the following four functions have been created.

1. `select_celcius`: This function split the string e.g. "61.58(16.4)" into two number 61.48 and 16.4 and select the *smallest* of the two. This ensure that metric represenation of the temperature is always selected i.e. celcius (except in the case when it is colder than -40 °C. But this is no worry for this sort of data).
1. `select_mm`: This function split the string e.g. "2.22(56)" into two number 2.22 and 56 and selects the *largest*. This ensure that metric represenation of the temperature is always selected i.e. minimeters. 
1. `ensure_numeric`: This function enures that the data is formatted as float.
1. `convert_to_metric`:  This function utilized the above functions to convert the entire dataframe into metric and numeric format.


In [24]:
def select_celcius(row):

    matric = []
    for imperial_metric in row:
        try:
            val1, val2 = imperial_metric.replace(")", "").split('(')
            
            # Remove symbol which look like a minus sign which is not not a minus sign
            val1, val2 = val1.replace('−', '-'), val2.replace('−', '-')
            # Remove commas which seperates tree zeros and convert to float 
            val1, val2 = float(val1.replace(',', '')), float(val2.replace(',', ''))

            matric.append(min(val1, val2))
        except ValueError:
            try:
                matric.append(float(imperial_metric))
            except ValueError:
                matric.append(None)
    return matric


def select_mm(row):

    matric = []
    for imperial_metric in row:
        try:
            val1, val2 = imperial_metric.replace(")", "").split('(')
            
            # Remove symbol which look like a minus sign which is not not a minus sign
            val1, val2 = val1.replace('−', '-'), val2.replace('−', '-')
            # Remove commas which seperates tree zeros and convert to float 
            val1, val2 = float(val1.replace(',', '')), float(val2.replace(',', ''))

            matric.append(max(val1, val2))
        except ValueError:
            try:
                matric.append(float(imperial_metric))
            except ValueError:
                matric.append(None)
    return matric


def ensure_numeric(row):
    metric = []
    for val in row:
        try:
            val = float(val.replace(',', ''))
            metric.append(val)
        except ValueError:
            matric.append(np.nan)
    
    return metric


def convert_to_metric(city_climate_df, function_dict):
    
    columns = city_climate_df.columns

    # Init data frame
    metric_city_climate_df = pd.DataFrame(columns = columns)

    climate_variables = []
    # Populate data frame
    for climate_var, row in city_climate_df.iterrows():
        climate_variables.append(climate_var)
        metric = function_dict[climate_var]( city_climate_df.loc[climate_var] )

        tmp = pd.DataFrame({key: [value] for key, value in zip(columns, metric)})
        metric_city_climate_df = metric_city_climate_df.append(tmp)

    metric_city_climate_df['Climate_variable'] = climate_variables
    metric_city_climate_df.set_index('Climate_variable', inplace = True)

    return metric_city_climate_df

Now all the function are utlized on the data for Austin. As can be seen below only the metric repreation is left. 

In [25]:
function_dict = {
    'Average high °C' : select_celcius,
    'Average low °C' : select_celcius,
    'Average precipitation mm' : select_mm,
    'Average precipitation days' : ensure_numeric,
    'Mean monthly sunshine hours' : ensure_numeric
}

convert_to_metric(
    city_climate_df = austin_tmp,
    function_dict = function_dict
)

Jan    Feb    Mar    Apr    May    Jun    Jul  \
Climate_variable                                                               
Average high °C               16.4   18.4   22.3   26.6   30.3   33.4   35.3   
Average low °C                 5.3    7.1   10.7   14.8   19.3   22.4   23.6   
Average precipitation mm      56.0   51.0   70.0   53.0  113.0  110.0   48.0   
Average precipitation days     7.4    7.4    9.2    7.1    8.9    7.7    5.4   
Mean monthly sunshine hours  163.8  169.3  205.9  205.8  227.1  285.5  317.2   

                               Aug    Sep    Oct    Nov    Dec    Year  
Climate_variable                                                        
Average high °C               36.1   32.5   27.7   21.9   17.1    26.6  
Average low °C                23.7   20.8   15.9   10.3    5.7    15.0  
Average precipitation mm      60.0   76.0   99.0   75.0   61.0   872.0  
Average precipitation days     4.9    6.7    7.5    7.5    7.8    87.5  
Mean monthly sunshine hours  297.9  233.8  215.6  168.3  153.5  2643.7

### Processing all cities

Now the selection of cliamte variables and keeping their numeric metric format will be performed on all cities. 

In [26]:
index = pd.Index(
    ['Average high °C', 'Average low °C','Average precipitation mm', 'Average precipitation days', 'Mean monthly sunshine hours'],
    dtype='object', name='Climate_variable')

# Init data frame
empty_df = pd.DataFrame({
    'Jan'   : np.repeat(np.nan, 5),
    'Feb'   : np.repeat(np.nan, 5),
    'Mar'   : np.repeat(np.nan, 5),
    'Apr'   : np.repeat(np.nan, 5),
    'May'   : np.repeat(np.nan, 5), 
    'Jun'   : np.repeat(np.nan, 5),
    'Jul'   : np.repeat(np.nan, 5),
    'Aug'   : np.repeat(np.nan, 5),
    'Sep'   : np.repeat(np.nan, 5),
    'Oct'   : np.repeat(np.nan, 5),
    'Nov'   : np.repeat(np.nan, 5),
    'Dec'   : np.repeat(np.nan, 5),
    'Year'  : np.repeat(np.nan, 5),
}, index = index)

cleaned_city_dict = dict()
for i, (city, city_climate_df) in enumerate(climate_data_dict.items()):

    if city_climate_df is None:
        # Add empty DF in case the city does not have climate table.
        cleaned_city_dict[city] = empty_df
        continue

    # Keep only selected columns and remname them:
    city_climate_df_update = keep_selected_varibles(
        city_climate_df = city_climate_df,
        rename_column = rename_column
    )

    # extract the metric representaiton
    metric_city_climate_df = convert_to_metric(
        city_climate_df = city_climate_df_update,
        function_dict = function_dict
    )

    # Add missing values
    metric_city_climate_df = metric_city_climate_df.append(empty_df)
    metric_city_climate_df = metric_city_climate_df[~metric_city_climate_df.index.duplicated(keep='first')]
    metric_city_climate_df = metric_city_climate_df.reindex(index)


    # Update the city dataframe such that it only keeps the choosen climate variables
    # and they are metric
    cleaned_city_dict[city] = metric_city_climate_df

#### Pivot the data for every city in one dataframe
Now every cliamte dataframes will be put togeter to form a big dataframe which contains the cliamte data for every city. 

In [27]:
def stack_climate_data(city_climate_df):
    stacked_city_climate_date = city_climate_df.stack().to_frame()

    # Fix index
    stacked_city_climate_date.index = stacked_city_climate_date.index.map('_'.join)

    # Return transposed version
    return stacked_city_climate_date.T


for i, city in enumerate(cleaned_city_dict):
    city_df = pd.DataFrame({
        'wiki_link' : [city]
    })
    climate_data = stack_climate_data(cleaned_city_dict[city])

    city_cliamte_df = pd.concat([city_df, climate_data], axis=1, join='inner')

    if i == 0:
        all_cities_climate_df = city_cliamte_df

    else:
        all_cities_climate_df = all_cities_climate_df.append(city_cliamte_df)


final_data = pd.merge(df_geomet, all_cities_climate_df, how = 'inner').drop(columns='wiki_link')

## Imputing the Final Data
Now all that is missing is imputing the missing data such that the data can be used for modelling. The imputing will be carried out with closest geographical neighbours scheme as doen with the original data (though now ignoring the typology as it should not have an impact on the geographics or weather).

The havesine is again applied to calculate the distance between every city.

In [28]:
# Haversine 
def haversine(point1, point2):
    # convert decimal degrees to radians
    lon1, lat1 = map(np.radians, point1)
    lon2, lat2 = map(np.radians, point2)

    # Deltas
    delta_lon = lon2 - lon1 
    delta_lat = lat2 - lat1 
    
    # haversine formula 
    a = np.sin(delta_lat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(delta_lon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    r = 6371 # Radius of earth in km
    return c * r

df_dist = pd.DataFrame(index = final_data.City, columns = final_data.City)
for City in df_dist.index:
    df_dist.loc[City] = {
        c : haversine(
                point1 = final_data[final_data['City'] == City][["Latitude", "Longitude"]].values[0], 
                point2 = final_data[final_data['City'] == c][["Latitude", "Longitude"]].values[0]
            ) 
        for c in final_data.City
    }

# Check numbers
df_dist.head()
    

City          Baltimore(MD) Melbourne   Niamey    Hanoi   Urumqi  \
City                                                               
Baltimore(MD)             0   14113.7  8902.46  19437.2  18228.9   
Melbourne           14113.7         0  13270.2   5347.1  6161.78   
Niamey              8902.46   13270.2        0  11520.5  9514.63   
Hanoi               19437.2    5347.1  11520.5        0   2041.6   
Urumqi              18228.9   6161.78  9514.63   2041.6        0   

City          Milwaukee(WI)    Paris Montevideo   Krakow Yekaterinburg  ...  \
City                                                                    ...   
Baltimore(MD)       1258.39  8801.07    3613.66  10760.2       15307.9  ...   
Melbourne           13937.3  10160.2    17652.5  8937.06       6427.59  ...   
Niamey              10041.1  3925.74    7808.45  4434.74       7462.55  ...   
Hanoi                 18464  11302.9    16921.9   9347.2       4799.54  ...   
Urumqi                19486  9449.35    16185.6  7494.64       2979.02  ...   

City              Wuxi   Harare Monterrey  Yerevan Liverpool   Durban  \
City                                                                    
Baltimore(MD)  18110.4  12587.7   2617.31  13468.4   8232.39  12843.3   
Melbourne      5569.27  12375.6   13338.9  8192.79   10087.4  12624.1   
Niamey         12966.5  4603.45     11363  5392.26   4469.24  5571.17   
Hanoi          1664.62  7973.65   17107.5  6742.38   11819.7  7743.81   
Urumqi         3662.13  6415.37   19147.3  4760.41   10046.6  6477.56   

City            Kumasi    Izmir   Ankara Cape Town  
City                                                
Baltimore(MD)  8577.55  11537.1    12172   11577.4  
Melbourne      14106.5  9443.12  8949.34   14057.1  
Niamey         865.047  3843.67  4389.08   5476.46  
Hanoi          11894.3  8680.03  8034.75   9015.44  
Urumqi         9949.07  6692.44  6056.85   7896.58  

[5 rows x 331 columns]

Based on the distances can the closest neighbour be found and with these neighbours the data can be imputed in both a weighted and unweighted fashion

In [29]:
def KNN_imputer(df, df_dist, k, Weighted = False):

    # Copy of df to return
    df_new = df.copy()

    for City in df.City:
        # Get columns where data is missing
        missing = df[df.City == City].isna().squeeze()[df[df.City == City].isna().squeeze()].index

        for col in missing:
            # Get the k closes cities
            k_closest_cities = df_dist.loc[[City]].T.sort_values(by = City, ascending=True).head(k+1)

            # exstract the k closest city names
            closest_cities = k_closest_cities.index[1:(k+1)].to_list()

            if Weighted:
                distances = df_dist.loc[[City]][closest_cities].values
                weights = distances/distances.sum()
                imputed = df_new[df_new['City'].isin(closest_cities)][missing].mul(weights[0], axis = 0).sum()

            else:
                imputed = df_new[df_new['City'].isin(closest_cities)][missing].mean()
        
            df_new.loc[(df_new['City'] == City), missing] = imputed.values

    return df_new

# Apply the imputer
df_final_imputed = KNN_imputer(
    df = final_data,
    df_dist = df_dist,
    k = 5,
    Weighted = True
)

It can be confirmed that there is now no missing values with cell below:

In [30]:
# Count number of NA
df_final_imputed.isna().sum().sum()

0

### Last fixing

As with the original data all column names are changed so they do not contain space:

In [31]:
#%% Add _ instead of space
df_final_imputed.columns = df_final_imputed.columns.str.replace(' ', '_')

The yearly columns in the data can be seen as linear combinations of the monthly variables - of this reason they are dropped from the data set. The same for the latitude and longitude as they are not needed anymore

In [32]:
# Drop yearly columns
df_final_imputed = df_final_imputed.drop(columns = ['Latitude', 'Longitude', 'Average_high_°C_Year','Average_low_°C_Year', 'Average_precipitation_mm_Year','Average_precipitation_days_Year', 'Mean_monthly_sunshine_hours_Year'])

And now it is saved to be used for analysis

In [33]:
# Save data frame
df_final_imputed.to_csv("../data/processed/GeoMet_imputed.csv")